crates/wasmtime/src/config.rs

```diff
@@ -1283,126 +1283,210 @@ impl Config {
+    /// Wasmtime.
+    ///
+    /// ## Default Value
     ///
-    /// Note that the pooling allocator can reduce the amount of memory needed
-    /// for pooling allocation by using memory protection; see
-    /// `PoolingAllocatorConfig::memory_protection_keys` for details.
-    pub fn static_memory_maximum_size(&mut self, max_size: u64) -> &mut Self {
-        self.tunables.static_memory_reservation = Some(max_size);
+    /// The default value for this property depends on the host platform. For
+    /// 64-bit platforms there's lots of address space available, so the default
+    /// configured here is 4GiB. When coupled with the default size of
+    /// [`Config::memory_guard_size`] this means that 32-bit WebAssembly linear
+    /// memories with 64KiB page sizes will skip almost all bounds checks by
+    /// default.
+    ///
+    /// For 32-bit platforms this value defaults to 10MiB. This means that
+    /// bounds checks will be required on 32-bit platforms.
+    pub fn memory_reservation(&mut self, bytes: u64) -> &mut Self {
+        self.tunables.memory_reservation = Some(bytes);
         self
     }
 
-    /// Indicates that the "static" style of memory should always be used.
+    /// Indicates that [`Config::memory_reservation`] is the maximal size of any
+    /// linear memory.
+    ///
+    /// WebAssembly linear memories either have a maximum size that's explicitly
+    /// listed in the type of a memory or inherently limited by the index type
+    /// of the memory (e.g. 4GiB for 32-bit linear memories). Memory growth can
+    /// be limited via [`Store::limiter`](crate::Store::limiter) but this option
+    /// serves as another "easy" means of limiting the growth of linear memory.
     ///
-    /// This configuration option enables selecting the "static" option for all
-    /// linear memories created within this `Config`. This means that all
-    /// memories will be allocated up-front and will never move. Additionally
-    /// this means that all memories are synthetically limited by the
-    /// [`Config::static_memory_maximum_size`] option, regardless of what the
-    /// actual maximum size is on the memory's original type.
+    /// When specified this option will automatically cap the size of linear
+    /// memories to the value of [`Config::memory_reservation`] meaning that
+    /// after initial allocation a linear memory is never allowed to move in
+    /// memory. See more information on [`Config::memory_reservation`] for
+    /// various effects on performance this can have.
     ///
-    /// For the difference between static and dynamic memories, see the
-    /// [`Config::static_memory_maximum_size`].
-    pub fn static_memory_forced(&mut self, force: bool) -> &mut Self {
-        self.tunables.static_memory_bound_is_maximum = Some(force);
+    /// The default value for this option is `false`.
+    pub fn memory_reservation_is_maximum(&mut self, enable: bool) -> &mut Self {
```

Comment:
I don't love the name `memory_reservation_is_maximum`. Here are a few alternative bike sheds:

* `memory_reservation_is_final`
* `memory_reservation_is_immovable`
* `memory_reservation_moves`
* `memory_reservation_relocates`
* `memory_reservation_reallocs`
* `memory_is_final`
* `memory_is_immovable`
* `memory_moves`
* `memory_relocates`
* `memory_reallocs`

I don't love any of these. Soft preference for `memory_moves`, I guess? Thoughts?

Reply:
Ok I've settled on `memory_may_move`, mind doing another pass over the docs I have for the method?